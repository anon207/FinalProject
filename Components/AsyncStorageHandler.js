import AsyncStorage from '@react-native-async-storage/async-storage';
import { useEffect } from 'react';

export const AsyncStorageHandler = ({ dataKey, data, setData }) => {
  useEffect(() => {
    const loadData = async () => {
      try {
        const storedData = await AsyncStorage.getItem(dataKey);
        if (storedData !== null) {
          setData(JSON.parse(storedData));
        }
      } catch (error) {
        console.error(`Error loading ${dataKey} data:`, error);
      }
    };

    loadData();
  }, [dataKey, setData]);

  useEffect(() => {
    const saveData = async () => {
      try {
        await AsyncStorage.setItem(dataKey, JSON.stringify(data));
      } catch (error) {
        console.error(`Error saving ${dataKey} data:`, error);
      }
    };

    saveData();
  }, [dataKey, data]);

  return null; 
};