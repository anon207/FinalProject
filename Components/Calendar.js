import { StyleSheet, ScrollView, } from 'react-native';
import { useState, React } from 'react';
import { HeaderChangeMonthDayRow } from './HeaderChangeMonthDayRow';
import { RenderCalendarGridRowsAndEvents } from './RenderCalendarGridRowsAndEvents';

export const Calendar = ({ filteredEvents, setFilteredEvents, favorites, setFavorites, navigation }) => {
  const [selectedDay, setSelectedDay] = useState(null);
  const [currentMonth, setCurrentMonth] = useState(0);

  const months = ['January', 'February', 'March', 'April', 'May', 'June', 'July', 'August', 'September', 'October', 'November', 'December'];
  const formattedMonth = currentMonth + 1 < 10 ? `0${currentMonth + 1}` : `${currentMonth + 1}`;
  const formattedDay = selectedDay < 10 ? `0${selectedDay}` : selectedDay;
  const fullDate = `2024-${formattedMonth}-${formattedDay}`;

  const eventsForDay = filteredEvents.filter(event => event.date === fullDate);

  const toggleEvents = day => setSelectedDay(selectedDay === day ? null : day);

  const toggleFavorite = async (event) => {
    const updatedEvents = filteredEvents.map(e => e.Id === event.Id ? { ...e, favorite: !e.favorite } : e);

    if (event.favorite) {
        setFavorites(prevFavorites => [...prevFavorites, event]);
    } else {
        setFavorites(prevFavorites => prevFavorites.filter(favorite => favorite.Id !== event.Id));
    }

    setFilteredEvents(updatedEvents);
  };

  const changeMonth = (increment) => {
    setCurrentMonth((prevMonth) => (prevMonth + increment + 12) % 12);
  };

  const renderCalendarRows = () => {
    const calendarRows = [];
    let currentRow = [];
    for (let i = 1; i <= daysInMonth(months[currentMonth]); i++) {
      currentRow.push(i);
      if (currentRow.length === 7 || i === daysInMonth(months[currentMonth])) {
        calendarRows.push([...currentRow]);
        currentRow = [];
      }
    }
    return calendarRows;
  };

  const daysInMonth = (month) => {
    if (['January', 'March', 'May', 'July', 'August', 'October', 'December'].includes(month)) return 31;
    if (['April', 'June', 'September', 'November'].includes(month)) return 30;
    return month === 'February' ? 29 : 0;
  };
  
  return (
    <ScrollView contentContainerStyle={CalendarStyles.calendar} testID='calendar-component'>
      <HeaderChangeMonthDayRow
        setFilteredEvents={setFilteredEvents}
        navigation={navigation}
        onForward={() => changeMonth(1)}
        onBackward={() => changeMonth(-1)}
        currMonth={months[currentMonth]}
      />
      <RenderCalendarGridRowsAndEvents 
        renderCalendarRows={renderCalendarRows}
        currentMonth={currentMonth}
        toggleEvents={toggleEvents}
        selectedDay={selectedDay}
        filteredEvents={filteredEvents}
        eventsForDay={eventsForDay}
        favorites={favorites}
        toggleFavorite={toggleFavorite}
      />
    </ScrollView>
  );
};

const CalendarStyles = StyleSheet.create({
    calendar: {
      flexDirection: 'column',
      flexWrap: 'wrap',
      paddingLeft: 2.5,
      alignItems: 'center',
    },
});