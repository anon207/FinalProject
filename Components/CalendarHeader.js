import { StyleSheet, View } from 'react-native';
import { React } from 'react';
import { EventFilter } from './EventFilter';
import { FavoriteList } from './FavoriteList';

export const CalendarHeader = ({ setFilteredEvents, navigation, favorites }) => {
    return(
      <View style={CalendarStyles.headerContainer}>
          <EventFilter setFilteredEvents={setFilteredEvents}/>
          <FavoriteList navigation={navigation} favorites={favorites}/>
      </View>
    );
  };

  export const CalendarStyles = StyleSheet.create({
    calendarGrid: {
      flexDirection: 'row',
      flexWrap: 'wrap',
      justifyContent: 'flex-start',
    },
    calendar: {
      flexDirection: 'column',
      flexWrap: 'wrap',
      paddingLeft: 2.5,
      alignItems: 'center',
    },
    EventDisplay: {
      width: 385,
      height: 100,
      borderWidth: 1,
      borderColor: 'lightgray',
      marginBottom: 20,
      alignItems: 'center',
      justifyContent: 'center',
    },
    firstEvent: {
      height: 100,
      borderWidth: 1,
      borderColor: 'lightgray',
      marginBottom: 20,
      marginTop: 20,
      alignItems: 'center',
      justifyContent: 'center',
    },
    homeStyle: {
      height: 100,
      width: 5,
      backgroundColor: 'maroon',
      alignSelf: 'flex-start',
      position: 'absolute',
      zIndex: 1,
    },
    bottomBar: {
      width: 385,
      height: 10,
      borderBottomColor: 'white',
      borderColor: 'lightgray',
      borderWidth: 1,
      position: 'absolute',
      bottom: 0
    },
    favButton: {
      height: 32.5,
      width: 65, 
      backgroundColor: 'maroon',
      justifyContent: 'center',
      alignItems: 'center',
      borderRadius: 5,
      position: 'absolute',
      right: 5,
      top: 5,
    },
    Remove: {
      height: 32.5,
      width: 65, 
      backgroundColor: 'rgba(143, 11, 11, 0.5)',
      justifyContent: 'center',
      alignItems: 'center',
      borderRadius: 5,
      position: 'absolute',
      right: 5,
      top: 5,
      borderWidth: 1,
      borderColor: 'maroon',
    },
    headerContainer: {
      flexDirection: 'row',
      justifyContent: 'space-around',
      alignItems: 'center',
      width: '100%', 
      marginVertical: 25,
    },
  });