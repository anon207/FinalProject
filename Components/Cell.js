import { StyleSheet, Text, View, Pressable } from 'react-native';
import React from 'react';

export const Cell = ({ day, toggleEvents, isSelected, filteredEvents, favorites }) => {
  let containsHome = false;
  for(const event of filteredEvents) {
    if(event.homeAway === 'Home') {
      containsHome = true;
    }
  }

  let containsFav = false;
  for(const event of filteredEvents) {
    if(favorites.some(favorite => favorite.Id === event.Id)) {
      containsFav = true;
    }
  }
console.log(favorites[0]);
  return(
    <Pressable onPress={() => toggleEvents(day)} testID={`day-${day}`}>
      {({ pressed }) => (
        <View style={[CellStyles.cell, (filteredEvents.length > 0 && isSelected) && CellStyles.selectedCell, (filteredEvents.length > 0 && pressed) && CellStyles.pressedCell, (containsFav) && CellStyles.favorited]}>
          <View style={CellStyles.cellDay}>
            <Text style={CellStyles.dayText}>{day}</Text>
          </View>
          <View style={CellStyles.eventPosition}>
            {(filteredEvents.length > 0) && (!isSelected) && 
              <Text style={CellStyles.evt}>{filteredEvents.length} Events</Text>
            }
            {(containsHome && !isSelected) &&
              <View style={CellStyles.HomeStyle} testID='home-marker'/>
            }
            {(isSelected && filteredEvents.length > 0) &&
            <View style={CellStyles.Xcontainer}>
              <View style={[CellStyles.line, CellStyles.lineDiagonalLeft]} />
              <View style={[CellStyles.line, CellStyles.lineDiagonalRight]} />
            </View>
            }
          </View>
        </View>
      )}
    </Pressable>
  );
};

const CellStyles = StyleSheet.create({
  favorited: {
    backgroundColor: 'maroon',
  },
  cell: {
    height: 100,
    width: 55,
    backgroundColor: 'white',
    borderWidth: 1,
    borderColor: 'lightgray',
  },
  selectedCell: {
    borderBottomColor: 'white'
  },
  pressedCell: {
    backgroundColor: 'rgba(173, 216, 230, 0.5)',
  },
  cellDay: {
    flex: 1,
    alignItems: 'center',
    paddingTop: 15,
  },
  dayText: {
    fontFamily: 'RobotoCondensed-Bold',
    fontWeight: 'bold',
    fontSize: 18,
  },
  eventPosition: {
    alignItems: 'center',
    paddingBottom: 15,
  },
  evt: {
    fontFamily: 'RobotoCondensed-Regular',
    fontWeight: 'bold',
    fontSize: 10,
    color: '#666666'
  },
  Xcontainer: {
    flexDirection: 'row',
    justifyContent: 'center',
    alignItems: 'center',
    paddingBottom: 25,
  },
  line: {
    position: 'absolute',
    backgroundColor: 'black',
  },
  lineDiagonalLeft: {
    width: 5,
    height: 18,
    borderRadius: 1,
    transform: [{ rotate: '45deg' }],
  },
  lineDiagonalRight: {
    width: 5,
    height: 18,
    borderRadius: 1,
    transform: [{ rotate: '-45deg' }],
  },
  HomeStyle: {
    width: 55,
    height: 7.5,
    backgroundColor: 'maroon',
    position: 'absolute',
    bottom: 0,
  }
});