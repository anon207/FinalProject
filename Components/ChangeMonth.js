import { StyleSheet, Text, View, Pressable, } from 'react-native';
import React from 'react';

export const ChangeMonth = ({ currMonth, onForward, onBackward }) => {
    return(
       <View style={ChangeMonthStyles.monthChange}>
         <Pressable onPress={onBackward}>
           <Text style={{ fontSize: 30, fontFamily: 'RobotoCondensed-Regular' }}>{'<< '}</Text>
         </Pressable>
         <Text style={{ fontSize: 30, fontFamily: 'RobotoCondensed-Regular' }}> {currMonth} </Text>
         <Pressable onPress={onForward}>
           <Text style={{ fontSize: 30, fontFamily: 'RobotoCondensed-Regular' }}>{' >>'}</Text>
         </Pressable>
       </View>
     );
   };

   const ChangeMonthStyles = StyleSheet.create({
    monthChange: {
      alignItems: 'center',
      justifyContent: 'center',
      flexDirection: 'row',
      paddingBottom: 10,
    },
  });