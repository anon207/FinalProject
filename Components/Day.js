import { StyleSheet, Text, View } from 'react-native';
import React from 'react';


export const Day = ({ day }) => {
    return(
      <View style={DayStyles.day}>
        <Text style={DayStyles.dayFont}>{day}</Text>
      </View>
    );
};

const DayStyles = StyleSheet.create({
    day: {
      flex: 1,
      alignItems: 'center',
      paddingVertical: 10,
    },
    dayFont: {
      fontFamily: 'RobotoCondensed-Bold',
      fontWeight: 'bold',
      fontSize: 18, 
      color: '#666666'
    },
});