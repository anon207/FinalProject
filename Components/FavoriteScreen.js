import { StyleSheet, Text, View, ScrollView, Pressable } from 'react-native';
import React, { useContext } from 'react';
import { FavoritesContext } from './FavoriteContext';

export const FavoritesScreen = () => {
  const { favorites, setFavorites, filteredEvents, setFilteredEvents } = useContext(FavoritesContext);

  const toggleFavorite = async (event) => {
    const updatedEvents = filteredEvents.map(e => e.Id === event.Id ? { ...e, favorite: false } : e);
    setFavorites(prevFavorites => prevFavorites.filter(favorite => favorite.Id !== event.Id));
    setFilteredEvents(updatedEvents);
  };

  return(
    <ScrollView contentContainerStyle={FavoriteScreenStyles.defualtView}>
      <Pressable onPress={() => setFavorites([])}>
        {({ pressed }) => (
          <View style={[FavoriteScreenStyles.removeAll, pressed && FavoriteScreenStyles.pressedEvt]}>
            <Text style={{color: 'white', fontFamily: 'RobotoCondensed-Regular'}}>Unfavorite all</Text>
          </View>
        )}
      </Pressable>
      {favorites.length === 0 &&
        <View style={FavoriteScreenStyles.noEvents}>
          <Text style={{fontFamily: 'RobotoCondensed-Regular', fontSize: 24}}>
            You have no favorited events!
          </Text>
        </View>
      }
      {favorites.map((event, eventIndex) => (
        <View key={eventIndex} style={[FavoriteScreenStyles.EventDisplay, (eventIndex === 0) && FavoriteScreenStyles.firstEvent]}>
          {event.homeAway === 'Home' &&
            <View style={FavoriteScreenStyles.homeStyle} testID={`event-${eventIndex}-home-marker`}/>
          }
          <Text style={{ fontFamily: 'RobotoCondensed-Bold' }}>{event.name}</Text>
          <Text style={{ fontFamily: 'RobotoCondensed-Bold' }}>{event.time}</Text>
          <Text style={{ fontFamily: 'RobotoCondensed-Bold' }}>{event.location}</Text>
          <View style={FavoriteScreenStyles.bottomBar} />
          <Pressable
            key={event.Id}
            style={( (favorites.some(favorite => favorite.Id === event.Id)) ) ? FavoriteScreenStyles.Remove : FavoriteScreenStyles.favButton}
            onPress={() => toggleFavorite(event)}
            testID={`event-${eventIndex}-fav-button`}
          >
            <Text style={{ color: 'white', fontSize: 10, fontFamily: 'RobotoCondensed-Regular' }}>
              {( (favorites.some(favorite => favorite.Id === event.Id)) ) ? 'Unfavorite' : 'Favorite'}
            </Text>
          </Pressable>
        </View>
      ))}
    </ScrollView>
  );
};

const FavoriteScreenStyles = StyleSheet.create({  
  pressedEvt: {
    backgroundColor: 'rgba(87, 0, 0, 1.0)',
  },
  noEvents: {
    width: 385,
    height: 800,
    alignItems: 'center'
  },
  removeAll: {
    width: 100,
    height: 50,
    backgroundColor: 'maroon',
    borderRadius: 5,
    justifyContent: 'center',
    alignItems: 'center',
    marginTop: 20,
    marginBottom: 20,
  },
  defualtView: {
    width: 390,
    backgroundColor: '#fff',
    alignItems: 'center',
    justifyContent: 'center',
  },
  EventDisplay: {
    width: 385,
    height: 100,
    borderWidth: 1,
    borderColor: 'lightgray',
    marginBottom: 20,
    alignItems: 'center',
    justifyContent: 'center',
  },
  firstEvent: {
    height: 100,
    borderWidth: 1,
    borderColor: 'lightgray', 
    marginBottom: 20,
    alignItems: 'center',
    justifyContent: 'center',
  },
  bottomBar: {
    width: 385,
    height: 10,
    borderBottomColor: 'white',
    borderColor: 'lightgray',
    borderWidth: 1,
    position: 'absolute',
    bottom: 0
  },
  homeStyle: {
    height: 100,
    width: 5,
    backgroundColor: 'maroon',
    alignSelf: 'flex-start',
    position: 'absolute',
    zIndex: 1,
  },
  favButton: {
    height: 32.5,
    width: 65, 
    backgroundColor: 'maroon',
    justifyContent: 'center',
    alignItems: 'center',
    borderRadius: 5,
    position: 'absolute',
    right: 5,
    top: 5,
  },
  Remove: {
    height: 32.5,
    width: 65, 
    backgroundColor: 'rgba(143, 11, 11, 0.5)',
    justifyContent: 'center',
    alignItems: 'center',
    borderRadius: 5,
    position: 'absolute',
    right: 5,
    top: 5,
    borderWidth: 1,
    borderColor: 'maroon',
  },    
});