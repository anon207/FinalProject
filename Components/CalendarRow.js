import { View, StyleSheet } from 'react-native';
import React from 'react';
import { Cell } from './Cell';

export const CalendarRow = ({ days, currentMonth, toggleEvents, selectedDay, filteredEvents}) => {
    return (
      <View style={CalendarRowStyles.calendarGrid}>
        {days.map(day => {
          const formattedMonth = currentMonth + 1 < 10 ? `0${currentMonth + 1}` : `${currentMonth + 1}`;
          const formattedDay = day < 10 ? `0${day}` : day;
          const fullDate = `2024-${formattedMonth}-${formattedDay}`;
  
          // Filter events for this specific day
          const eventsForDay = filteredEvents.filter(event => event.date === fullDate);
  
          return (
            <Cell 
              key={day}
              day={day}
              currentMonth={currentMonth}
              toggleEvents={toggleEvents}
              isSelected={selectedDay === day}
              filteredEvents={eventsForDay} // Passing filtered events specific to this day
            />
          );
        })}
      </View>
    );
};

const CalendarRowStyles = StyleSheet.create({
    calendarGrid: {
      flexDirection: 'row',
      flexWrap: 'wrap',
      justifyContent: 'flex-start',
    },
});