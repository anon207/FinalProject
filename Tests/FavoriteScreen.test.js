import React from 'react';
import { render, fireEvent } from '@testing-library/react-native';
import { FavoritesScreen } from '../Components/FavoriteScreen';
import { FavoritesContext } from '../Components/FavoriteContext';

describe('FavoritesScreen', () => {
  test('renders correctly', () => {
    const { toJSON } = render(
      <FavoritesContext.Provider value={{
        favorites: [],
        setFavorites: jest.fn(),
      }}>
        <FavoritesScreen favorites={[{
        name: "Test Event",
        time: "12:00",
        location: "Test Location",
        homeAway: "Home",
        Id: 0,
        date: "2024-04-01"
      }]}/>
      </FavoritesContext.Provider>
    );
    expect(toJSON()).toMatchSnapshot();
  });

  test('unfavoriting all events', () => {
    const { getByText, queryByText } = render(
      <FavoritesContext.Provider value={{
        favorites: [],
        setFavorites: jest.fn(),
      }}>
        <FavoritesScreen favorites={[{
        name: "Test Event",
        time: "12:00",
        location: "Test Location",
        homeAway: "Home",
        Id: 0,
        date: "2024-04-01"
      }]}/>
      </FavoritesContext.Provider>
    );
    fireEvent.press(getByText('Unfavorite all'));
    expect(queryByText('You have no favorited events!')).toBeTruthy();
  });

  test('toggling individual event favoriting/unfavoriting', () => {
    const mockFavorites = [
      {
        name: "Test Event",
        time: "12:00",
        location: "Test Location",
        homeAway: "Home",
        Id: 0,
        date: "2024-04-01"
      }
    ];
    const { getByText, queryByTestId } = render(
      <FavoritesContext.Provider value={{
        favorites: mockFavorites,
        setFavorites: jest.fn(),
      }}>
        <FavoritesScreen favorites={mockFavorites} />
      </FavoritesContext.Provider>
    );
    const favoriteButton = getByText('Unfavorite');
    fireEvent.press(favoriteButton);
    expect(queryByTestId('event-0-fav-button')).toBeTruthy();
  });
});