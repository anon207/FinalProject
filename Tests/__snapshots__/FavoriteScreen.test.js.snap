// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FavoritesScreen renders correctly 1`] = `
<RCTScrollView
  contentContainerStyle={
    {
      "alignItems": "center",
      "backgroundColor": "#fff",
      "justifyContent": "center",
      "width": 390,
    }
  }
>
  <View>
    <View
      accessibilityState={
        {
          "busy": undefined,
          "checked": undefined,
          "disabled": undefined,
          "expanded": undefined,
          "selected": undefined,
        }
      }
      accessibilityValue={
        {
          "max": undefined,
          "min": undefined,
          "now": undefined,
          "text": undefined,
        }
      }
      accessible={true}
      collapsable={false}
      focusable={true}
      onBlur={[Function]}
      onClick={[Function]}
      onFocus={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
    >
      <View
        style={
          [
            {
              "alignItems": "center",
              "backgroundColor": "maroon",
              "borderRadius": 5,
              "height": 50,
              "justifyContent": "center",
              "marginBottom": 20,
              "marginTop": 20,
              "width": 100,
            },
            false,
          ]
        }
      >
        <Text
          style={
            {
              "color": "white",
              "fontFamily": "RobotoCondensed-Regular",
            }
          }
        >
          Unfavorite all
        </Text>
      </View>
    </View>
    <View
      style={
        {
          "alignItems": "center",
          "height": 800,
          "width": 385,
        }
      }
    >
      <Text
        style={
          {
            "fontFamily": "RobotoCondensed-Regular",
            "fontSize": 24,
          }
        }
      >
        You have no favorited events!
      </Text>
    </View>
  </View>
</RCTScrollView>
`;
